test/boost/secondary_index_test.cc

```diff
@@ -1575,6 +1575,228 @@ SEASTAR_TEST_CASE(test_token_order) {
+
+        auto equal_range = [](int64_t value) { return format("token(pk) = {}", value); };
+
+        auto do_tests = [&] {
+            assert_that(q(inclusive_inclusive_range(testset_tokens[1], testset_tokens[5]))).is_rows().with_rows(get_result_rows(1, 5));
+            assert_that(q(inclusive_inclusive_range(testset_tokens[1], testset_tokens[5] - 1))).is_rows().with_rows(get_result_rows(1, 4));
+            assert_that(q(inclusive_inclusive_range(testset_tokens[1], testset_tokens[5] + 1))).is_rows().with_rows(get_result_rows(1, 5));
+            assert_that(q(inclusive_inclusive_range(testset_tokens[1] + 1, testset_tokens[5]))).is_rows().with_rows(get_result_rows(2, 5));
+
+            assert_that(q(exclusive_inclusive_range(testset_tokens[1], testset_tokens[4]))).is_rows().with_rows(get_result_rows(2, 4));
+
+            assert_that(q(inclusive_exclusive_range(testset_tokens[1], testset_tokens[4]))).is_rows().with_rows(get_result_rows(1, 3));
+
+            assert_that(q(exclusive_exclusive_range(testset_tokens[1], testset_tokens[4]))).is_rows().with_rows(get_result_rows(2, 3));
+
+            assert_that(q(inclusive_infinity_range(testset_tokens[3]))).is_rows().with_rows(get_result_rows(3, testset_pks.size() - 1));
+
+            assert_that(q(exclusive_infinity_range(testset_tokens[3]))).is_rows().with_rows(get_result_rows(4, testset_pks.size() - 1));
+
+            assert_that(q(infinity_inclusive_range(testset_tokens[3]))).is_rows().with_rows(get_result_rows(0, 3));
+
+            assert_that(q(infinity_exclusive_range(testset_tokens[3]))).is_rows().with_rows(get_result_rows(0, 2));
+            
+            assert_that(q("token(pk) < 0")).is_rows().with_rows(get_result_rows(0, 4));
+
+            assert_that(q("token(pk) > 0")).is_rows().with_rows(get_result_rows(5, testset_pks.size() - 1));
+
+            assert_that(q(equal_range(testset_tokens[3]))).is_rows().with_rows(get_result_rows(3, 3));
+
+            // empty range
+            assert_that(q("token(pk) < 5 AND token(pk) > 100")).is_rows().with_size(0);
+
+            // prepared statement
+            auto prepared_id = e.prepare(get_query("token(pk) >= ? AND token(pk) <= ?")).get0();
+            auto msg = e.execute_prepared(prepared_id, {
+                cql3::raw_value::make_value(long_type->decompose(testset_tokens[1])), cql3::raw_value::make_value(long_type->decompose(testset_tokens[5]))
+            }).get0();
+            assert_that(msg).is_rows().with_rows(get_result_rows(1, 5));
+
+            msg = e.execute_prepared(prepared_id, {
+                cql3::raw_value::make_value(long_type->decompose(testset_tokens[2])), cql3::raw_value::make_value(long_type->decompose(testset_tokens[6]))
+            }).get0();
+            assert_that(msg).is_rows().with_rows(get_result_rows(2, 6));
+        };
+
+        eventually(do_tests);
+
+        cquery_nofail(e, "CREATE INDEX t_global ON t(v)");
+
+        eventually(do_tests);
```

Comment:
A comment you can ignore:
While eventually() works here, as the index is eventually built, an alternative approach I always prefer is to wait until the index is built (we have a way to check this explicitly, but there is also an ugly way for small tables by reading the tables content and then trying to look it up with the index and seeing if you find it).
The benefit of this approach is that if you have a bug (or while you're developing the test) you'll see an immediate failure. With "eventually", you will need to wait until it retries and retries for a long time, failing every time.

Reply:
Trying to detect when index contains all rows gets a bit ugly because:
* each index is on a separate column
* indexed column has multiple different values
* there are local and global indexes

On a happy path `eventually` shouldn't add much time, maybe it could stay.

Reply:
Yes, it's perfectly fine for an already working tests. I don't like to *write* tests this way because it takes a very long time to see them fail. I love writing tests when I can see the test's results a fraction of the second after I saved a python file :-)